Core/MIPS/IR/IRCompBranch.cpp

```diff
@@ -0,0 +1,379 @@
+	case 17: BranchRSZeroComp(op, IRComparison::Less, true, false);  break; //R(MIPS_REG_RA) = PC + 8; if ((s32)R(rs) >= 0) DelayBranchTo(addr); else PC += 4; break;//bgezal
+	case 18: BranchRSZeroComp(op, IRComparison::GreaterEqual, true, true);  break;  //R(MIPS_REG_RA) = PC + 8; if ((s32)R(rs) <  0) DelayBranchTo(addr); else SkipLikely(); break;//bltzall
+	case 19: BranchRSZeroComp(op, IRComparison::Less, true, true);   break; //R(MIPS_REG_RA) = PC + 8; if ((s32)R(rs) >= 0) DelayBranchTo(addr); else SkipLikely(); break;//bgezall
+	default:
+		_dbg_assert_msg_(CPU,0,"Trying to compile instruction that can't be compiled");
+		break;
+	}
+}
+
+// If likely is set, discard the branch slot if NOT taken.
+void IRJit::BranchFPFlag(MIPSOpcode op, IRComparison cc, bool likely) {
+	if (js.inDelaySlot) {
+		ERROR_LOG_REPORT(JIT, "Branch in FPFlag delay slot at %08x in block starting at %08x", GetCompilerPC(), js.blockStart);
+		return;
+	}
+	int offset = _IMM16 << 2;
+	u32 targetAddr = GetCompilerPC() + offset + 4;
+
+	MIPSOpcode delaySlotOp = GetOffsetInstruction(1);
+	ir.Write(IROp::FpCondToReg, IRTEMP_LHS);
+	if (!likely)
+		CompileDelaySlot();
+
+	ir.Write(IROp::Downcount, 0, js.downcountAmount & 0xFF, js.downcountAmount >> 8);
+
+	FlushAll();
+	// Not taken
+	ir.Write(ComparisonToExit(cc), ir.AddConstant(GetCompilerPC() + 8), IRTEMP_LHS, 0);
+	// Taken
+	if (likely)
+		CompileDelaySlot();
+	FlushAll();
+	ir.Write(IROp::ExitToConst, ir.AddConstant(targetAddr));
+	js.compiling = false;
+}
+
+void IRJit::Comp_FPUBranch(MIPSOpcode op) {
+	switch((op >> 16) & 0x1f) {
+	case 0:	BranchFPFlag(op, IRComparison::NotEqual, false); break;  // bc1f
+	case 1: BranchFPFlag(op, IRComparison::Equal, false); break;  // bc1t
+	case 2: BranchFPFlag(op, IRComparison::NotEqual, true);  break;  // bc1fl
+	case 3: BranchFPFlag(op, IRComparison::Equal, true);  break;  // bc1tl
+	default:
+		_dbg_assert_msg_(CPU, 0, "Trying to interpret instruction that can't be interpreted");
+		break;
+	}
+}
+
+// If likely is set, discard the branch slot if NOT taken.
+void IRJit::BranchVFPUFlag(MIPSOpcode op, IRComparison cc, bool likely) {
+	if (js.inDelaySlot) {
+		ERROR_LOG_REPORT(JIT, "Branch in VFPU delay slot at %08x in block starting at %08x", GetCompilerPC(), js.blockStart);
+		return;
+	}
+	int offset = _IMM16 << 2;
+	u32 targetAddr = GetCompilerPC() + offset + 4;
+
+	MIPSOpcode delaySlotOp = GetOffsetInstruction(1);
+	logBlocks = 1;
+	ir.Write(IROp::VfpuCtrlToReg, IRTEMP_LHS, VFPU_CTRL_CC);
+
+	ir.Write(IROp::Downcount, 0, js.downcountAmount & 0xFF, js.downcountAmount >> 8);
+
+	// Sometimes there's a VFPU branch in a delay slot (Disgaea 2: Dark Hero Days, Zettai Hero Project, La Pucelle)
+	// The behavior is undefined - the CPU may take the second branch even if the first one passes.
+	// However, it does consistently try each branch, which these games seem to expect.
+	bool delaySlotIsBranch = MIPSCodeUtils::IsVFPUBranch(delaySlotOp);
+	if (!likely)
```

Comment:
`&& !delaySlotIsBranch`, right?  That's the point of going to + 4 in that case.

-[Unknown]